Objects/typeobject.c

```diff
@@ -7858,6 +7899,9 @@ type_ready_fill_dict(PyTypeObject *type)
     if (type_dict_set_doc(type) < 0) {
         return -1;
     }
+    if (cleanup_tp_dict(type) < 0) {
```

Comment:
Instead of adding a new helper function, I'd rather we fix the logic in the other functions called here in `type_ready_fill_dict()`.  It may make sense to merge this change first though.  (We have to factor in the risks of backporting any changes.)

Reply:
We should also add a test.  Something based on your reproducer, like the following, should work (in Lib/test/test_types.py):

```python
class SubinterpreterTests(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        global interpreters
        try:
            from test.support import interpreters
        except ModuleNotFoundError:
            raise unittest.Skip('subinterpreters required')
        import test.support.interpreters.channels

    @cpython_only
    def test_slot_wrappers(self):
        rch, sch = interpreters.channels.create()

        # For now it's sufficient to check int.__str__.
        # See https://github.com/python/cpython/issues/117482
        # and https://github.com/python/cpython/pull/117660.
        script = '''if True:
            text = repr(int.__str__)
            sch.send_nowait(text)
            '''

        exec(script)
        expected = rch.recv()

        interp = interpreters.create()
        interp.exec('from test.support import interpreters')
        interp.prepare_main(sch=sch)
        interp.exec(script)
        results = rch.recv()

        self.assertEqual(results, expected)
```

Reply:
Thanks for the suggestion. I added the tests where you proposed and confirmed that it's working as expected.